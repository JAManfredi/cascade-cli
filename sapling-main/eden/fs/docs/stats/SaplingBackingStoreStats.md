# SaplingBackingStoreStats

1. `Duration get{xxx}{"store.sapling.get_{xxx}_us"}` :

Duration of the whole get xxx (blob, blobmetadata, tree, treemetadata)
SaplingBackingStore::get{xxx} in Microsecond. This includes looking in local
first then if not found prepare the request, enqueue the request and then mark
it as finished when it is fulfilled.

1. `Duration fetch{xxx}{"store.sapling.fetch_{xxx}_us"}` :

Duration of fetching xxx (blob, blobmetadata, tree, treemetadata) requests from
the network in Microsecond.

1. `Duration getRootTree{"store.sapling.get_root_tree_us"}` :

Duration of getting a Root Tree from the Backing Store in Microsecond.

1. `Duration importManifestForRoot{"store.sapling.import_manifest_for_root_us"}`
   :

Duration of getting a manifest for Root from the Backing Store in Microsecond.

1. `Counter fetch{xxx}Local{"store.sapling.fetch_{xxx}_local"}` :

Number of xxx (blob, blobmetadata, tree, treemetadata) fetching locally from
hgcache

1. `Counter fetch{xxx}Success{"store.sapling.fetch_{xxx}_success"}` :

Number of xxx (blob, blobmetadata, tree, treemetadata) that fetch successfully
in the first try. (It could be local or remote)

1. `Counter fetch{xxx}Failure{"store.sapling.fetch_{xxx}_failure"}` :

Number of xxx (blob, blobmetadata, tree, treemetadata) that failed after retry.

1. `Counter fetch{xxx}RetrySuccess{"store.sapling.fetch_{xxx}_retry_success"}` :

Number of xxx (blob, tree) that fetch successfully in the retry. (It could be
local or remote)

1. `Counter getRootTreeLocal{"store.sapling.get_root_tree_local"}` :

Number of root trees fetching locally from Cache

1. `Counter getRootTreeSuccess{"store.sapling.get_root_tree_success"}` :

Number of root trees that fetch successfully in the first try. (It could be
local or remote)

1. `Counter getRootTreeFailure{"store.sapling.get_root_tree_failure"}` :

Number of root trees that failed after retry.

1. `Counter getRootTreeRetrySuccess{"store.sapling.get_root_tree_retry_success"}`
   :

Number of root trees that fetch successfully in the retry. (It could be local or
remote)

1. `Counter importManifestForRootLocal{ "store.sapling.import_manifest_for_root_local"}`
   :

Number of manifest for root fetching locally from Cache

1. `Counter importManifestForRootSuccess{"Store.sapling.import_manifest_for_root_success"}`
   :

Number of manifest for root that fetch successfully in the first try. (It could
be local or remote)

1. `Counter importManifestForRootFailure{"Store.sapling.import_manifest_for_root_failure"}`
   :

Number of manifest for root that failed after retry.

1. `Counter importManifestForRootRetrySuccess{"Store.sapling.import_manifest_for_root_retry_success"}`
   :

Number of manifests for root that fetch successfully in the retry. (It could be
local or remote)

1. `Duration prefetchBlob{"store.sapling.prefetch_blob_us"}` :

Duration of prefetching Blobs requests from BackingStore.

1. `Counter prefetchBlobLocal{"store.sapling.prefetch_blob_local"}` :

Number of Blobs prefetching locally from Cache

1. `Counter prefetchBlobSuccess{"store.sapling.prefetch_blob_success"}` :

Number of Blobs that prefetch successfully in the first try. (It could be local
or remote)

1. `Counter prefetchBlobFailure{"store.sapling.prefetch_blob_failure"}` :

Number of Blobs that failed in the prefetch after retry.

1. `Counter prefetchBlobRetrySuccess{"store.sapling.prefetch_blob_retry_success"}`
   :

Number of Blobs that prefetch successfully in the retry. (It could be local or
remote)

1. `Counter loadProxyHash{"store.sapling.load_proxy_hash"}` :

Count the number of times that a proxy hash gets loaded.
