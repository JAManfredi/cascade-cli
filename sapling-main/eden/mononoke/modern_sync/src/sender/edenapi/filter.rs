/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::HashMap;
use std::sync::Arc;

use anyhow::Ok;
use anyhow::Result;
use async_trait::async_trait;
use edenapi_types::AnyFileContentId;
use mercurial_types::HgChangesetId;
use mercurial_types::HgFileNodeId;
use mercurial_types::HgManifestId;
use mercurial_types::blobs::HgBlobChangeset;
use minibytes::Bytes;
use mononoke_types::BonsaiChangeset;
use mononoke_types::ChangesetId;

use crate::sender::edenapi::EdenapiSender;

#[derive(Eq, Hash, PartialEq)]
pub enum MethodFilter {
    UploadContents,
    UploadTrees,
    UploadFilenodes,
    SetBookmark,
    UploadIdenticalChangeset,
    FilterExistingCommits,
    ReadBookmark,
}

pub struct FilterEdenapiSender {
    inner: Arc<dyn EdenapiSender + Send + Sync>,
    allowed: HashMap<MethodFilter, bool>,
}

impl FilterEdenapiSender {
    pub fn new(
        inner: Arc<dyn EdenapiSender + Send + Sync>,
        allowed: HashMap<MethodFilter, bool>,
    ) -> Self {
        Self { inner, allowed }
    }
}

#[async_trait]
impl EdenapiSender for FilterEdenapiSender {
    async fn upload_contents(&self, contents: Vec<(AnyFileContentId, Bytes)>) -> Result<()> {
        if self
            .allowed
            .get(&MethodFilter::UploadContents)
            .map_or(false, |v| *v)
        {
            return self.inner.upload_contents(contents).await;
        } else {
            Ok(())
        }
    }

    async fn upload_trees(&self, trees: Vec<HgManifestId>) -> Result<()> {
        if self
            .allowed
            .get(&MethodFilter::UploadTrees)
            .map_or(false, |v| *v)
        {
            return self.inner.upload_trees(trees).await;
        } else {
            Ok(())
        }
    }

    async fn upload_filenodes(&self, fn_ids: Vec<HgFileNodeId>) -> Result<()> {
        if self
            .allowed
            .get(&MethodFilter::UploadFilenodes)
            .map_or(false, |v| *v)
        {
            return self.inner.upload_filenodes(fn_ids).await;
        } else {
            Ok(())
        }
    }

    async fn set_bookmark(
        &self,
        bookmark: String,
        from: Option<HgChangesetId>,
        to: Option<HgChangesetId>,
    ) -> Result<()> {
        if self
            .allowed
            .get(&MethodFilter::SetBookmark)
            .map_or(false, |v| *v)
        {
            return self.inner.set_bookmark(bookmark, from, to).await;
        } else {
            Ok(())
        }
    }

    async fn upload_identical_changeset(
        &self,
        css: Vec<(HgBlobChangeset, BonsaiChangeset)>,
    ) -> Result<()> {
        if self
            .allowed
            .get(&MethodFilter::UploadIdenticalChangeset)
            .map_or(false, |v| *v)
        {
            return self.inner.upload_identical_changeset(css).await;
        } else {
            Ok(())
        }
    }

    async fn filter_existing_commits(
        &self,
        ids: Vec<(HgChangesetId, ChangesetId)>,
    ) -> Result<Vec<ChangesetId>> {
        if self
            .allowed
            .get(&MethodFilter::FilterExistingCommits)
            .map_or(false, |v| *v)
        {
            return self.inner.filter_existing_commits(ids).await;
        } else {
            Ok(ids.into_iter().map(|(_, cs_id)| cs_id).collect())
        }
    }

    async fn read_bookmark(&self, bookmark: String) -> Result<Option<HgChangesetId>> {
        if self
            .allowed
            .get(&MethodFilter::ReadBookmark)
            .map_or(false, |v| *v)
        {
            return self.inner.read_bookmark(bookmark).await;
        } else {
            Ok(None)
        }
    }
}
