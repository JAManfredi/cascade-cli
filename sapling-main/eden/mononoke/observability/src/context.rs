/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::sync::Arc;
use std::sync::Mutex;

use anyhow::Error;
use cached_config::ConfigHandle;
use cached_config::ConfigStore;

use crate::config::ObservabilityConfig;
use crate::config::ScubaVerbosityLevel;
use crate::scuba::ScubaLoggingDecisionFields;
use crate::scuba::should_log_scuba_sample;

const CONFIGERATOR_OBSERVABILITY_CONFIG: &str = "scm/mononoke/observability/observability_config";

struct CfgrObservabilityContextInner {
    config_handle: ConfigHandle<ObservabilityConfig>,
}

impl CfgrObservabilityContextInner {
    fn new(config_store: &ConfigStore) -> Result<Self, Error> {
        let config_handle = config_store
            .get_config_handle_DEPRECATED(CONFIGERATOR_OBSERVABILITY_CONFIG.to_string())?;

        Ok(Self { config_handle })
    }

    fn should_log_scuba_sample(
        &self,
        verbosity_level: ScubaVerbosityLevel,
        logging_decision_fields: ScubaLoggingDecisionFields,
    ) -> bool {
        let config = self.config_handle.get();
        let scuba_config = &config.scuba_config;
        should_log_scuba_sample(verbosity_level, scuba_config, logging_decision_fields)
    }
}

/// A modifiable struct to be used in
/// the unit tests
pub struct TestObservabilityContextInner;

impl TestObservabilityContextInner {
    pub fn new() -> Arc<Mutex<Self>> {
        Arc::new(Mutex::new(Self))
    }

    fn should_log_scuba_sample(
        &self,
        _verbosity_level: ScubaVerbosityLevel,
        _logging_decision_fields: ScubaLoggingDecisionFields,
    ) -> bool {
        true
    }
}

/// A static `ObservabilityContext` to represent
/// traditional behavior with predefined log levels
#[derive(Debug, Clone)]
#[allow(unused)]
pub struct StaticObservabilityContextInner;

impl StaticObservabilityContextInner {
    fn new() -> Self {
        Self
    }

    fn should_log_scuba_sample(
        &self,
        verbosity_level: ScubaVerbosityLevel,
        _logging_decision_fields: ScubaLoggingDecisionFields,
    ) -> bool {
        verbosity_level == ScubaVerbosityLevel::Normal
    }
}

#[derive(Clone)]
enum ObservabilityContextInner {
    Dynamic(Arc<CfgrObservabilityContextInner>),
    Static(StaticObservabilityContextInner),
    Test(Arc<Mutex<TestObservabilityContextInner>>),
}

impl ObservabilityContextInner {
    fn new(config_store: &ConfigStore) -> Result<Self, Error> {
        Ok(Self::Dynamic(Arc::new(CfgrObservabilityContextInner::new(
            config_store,
        )?)))
    }

    fn new_static() -> Self {
        Self::Static(StaticObservabilityContextInner::new())
    }

    fn new_test(inner: Arc<Mutex<TestObservabilityContextInner>>) -> Self {
        Self::Test(inner)
    }

    pub fn should_log_scuba_sample(
        &self,
        verbosity_level: ScubaVerbosityLevel,
        logging_decision_fields: ScubaLoggingDecisionFields,
    ) -> bool {
        match self {
            Self::Dynamic(octx) => {
                octx.should_log_scuba_sample(verbosity_level, logging_decision_fields)
            }
            Self::Static(octx) => {
                octx.should_log_scuba_sample(verbosity_level, logging_decision_fields)
            }
            Self::Test(octx) => octx
                .lock()
                .expect("poiosoned lock")
                .should_log_scuba_sample(verbosity_level, logging_decision_fields),
        }
    }
}

#[derive(Clone)]
pub struct ObservabilityContext {
    inner: ObservabilityContextInner,
}

impl ObservabilityContext {
    pub fn new(config_store: &ConfigStore) -> Result<Self, Error> {
        Ok(Self {
            inner: ObservabilityContextInner::new(config_store)?,
        })
    }

    pub fn new_test(inner: Arc<Mutex<TestObservabilityContextInner>>) -> Self {
        Self {
            inner: ObservabilityContextInner::new_test(inner),
        }
    }

    pub fn new_static() -> Self {
        Self {
            inner: ObservabilityContextInner::new_static(),
        }
    }

    pub fn should_log_scuba_sample(
        &self,
        verbosity_level: ScubaVerbosityLevel,
        logging_decision_fields: ScubaLoggingDecisionFields,
    ) -> bool {
        self.inner
            .should_log_scuba_sample(verbosity_level, logging_decision_fields)
    }
}
