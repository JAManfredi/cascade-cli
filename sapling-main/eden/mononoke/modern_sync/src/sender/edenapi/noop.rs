/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use anyhow::Result;
use async_trait::async_trait;
use edenapi_types::AnyFileContentId;
use mercurial_types::HgChangesetId;
use mercurial_types::HgFileNodeId;
use mercurial_types::HgManifestId;
use mercurial_types::blobs::HgBlobChangeset;
use minibytes::Bytes;
use mononoke_types::BonsaiChangeset;
use mononoke_types::ChangesetId;

use crate::sender::edenapi::EdenapiSender;

#[derive(Default)]
pub struct NoopEdenapiSender {}

#[async_trait]
impl EdenapiSender for NoopEdenapiSender {
    async fn upload_contents(&self, _contents: Vec<(AnyFileContentId, Bytes)>) -> Result<()> {
        Ok(())
    }

    async fn upload_trees(&self, _trees: Vec<HgManifestId>) -> Result<()> {
        Ok(())
    }

    async fn upload_filenodes(&self, _fn_ids: Vec<HgFileNodeId>) -> Result<()> {
        Ok(())
    }

    async fn set_bookmark(
        &self,
        _bookmark: String,
        _from: Option<HgChangesetId>,
        _to: Option<HgChangesetId>,
    ) -> Result<()> {
        Ok(())
    }

    async fn upload_identical_changeset(
        &self,
        _css: Vec<(HgBlobChangeset, BonsaiChangeset)>,
    ) -> Result<()> {
        return Ok(());
    }

    async fn filter_existing_commits(
        &self,
        ids: Vec<(HgChangesetId, ChangesetId)>,
    ) -> Result<Vec<ChangesetId>> {
        Ok(ids.iter().map(|id| id.1).collect())
    }

    async fn read_bookmark(&self, _bookmark: String) -> Result<Option<HgChangesetId>> {
        Ok(None)
    }
}
