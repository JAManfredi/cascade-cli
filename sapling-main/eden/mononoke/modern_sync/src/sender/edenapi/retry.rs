/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::sync::Arc;
use std::time::Duration;

use anyhow::Result;
use async_trait::async_trait;
use edenapi_types::AnyFileContentId;
use futures::FutureExt;
use futures::future::BoxFuture;
use mercurial_types::HgChangesetId;
use mercurial_types::HgFileNodeId;
use mercurial_types::HgManifestId;
use mercurial_types::blobs::HgBlobChangeset;
use minibytes::Bytes;
use mononoke_types::BonsaiChangeset;
use mononoke_types::ChangesetId;

use crate::sender::edenapi::EdenapiSender;

const MAX_RETRIES: usize = 3;

pub struct RetryEdenapiSender {
    inner: Arc<dyn EdenapiSender + Send + Sync>,
}

impl RetryEdenapiSender {
    pub fn new(inner: Arc<dyn EdenapiSender + Send + Sync>) -> Self {
        Self { inner }
    }

    async fn with_retry<'t, T>(
        &'t self,
        func: impl Fn(&'t Self) -> BoxFuture<'t, Result<T>>,
    ) -> Result<T> {
        let retry_count = MAX_RETRIES;
        with_retry(retry_count, || func(self)).await
    }
}

#[async_trait]
impl EdenapiSender for RetryEdenapiSender {
    async fn upload_contents(&self, contents: Vec<(AnyFileContentId, Bytes)>) -> Result<()> {
        self.with_retry(|this| this.inner.upload_contents(contents.clone()).boxed())
            .await
    }

    async fn upload_trees(&self, trees: Vec<HgManifestId>) -> Result<()> {
        self.with_retry(|this| this.inner.upload_trees(trees.clone()).boxed())
            .await
    }

    async fn upload_filenodes(&self, fn_ids: Vec<HgFileNodeId>) -> Result<()> {
        self.with_retry(|this| this.inner.upload_filenodes(fn_ids.clone()).boxed())
            .await
    }

    async fn set_bookmark(
        &self,
        bookmark: String,
        from: Option<HgChangesetId>,
        to: Option<HgChangesetId>,
    ) -> Result<()> {
        self.inner.set_bookmark(bookmark, from, to).await
    }

    async fn upload_identical_changeset(
        &self,
        css: Vec<(HgBlobChangeset, BonsaiChangeset)>,
    ) -> Result<()> {
        self.with_retry(|this| this.inner.upload_identical_changeset(css.clone()).boxed())
            .await
    }

    async fn filter_existing_commits(
        &self,
        ids: Vec<(HgChangesetId, ChangesetId)>,
    ) -> Result<Vec<ChangesetId>> {
        self.inner.filter_existing_commits(ids).await
    }

    async fn read_bookmark(&self, bookmark: String) -> Result<Option<HgChangesetId>> {
        self.inner.read_bookmark(bookmark).await
    }
}

async fn with_retry<'t, T>(
    max_retry_count: usize,
    func: impl Fn() -> BoxFuture<'t, Result<T>>,
) -> Result<T> {
    let mut attempt = 0usize;
    loop {
        let result = func().await;
        if attempt >= max_retry_count {
            return result;
        }
        match result {
            Ok(result) => return Ok(result),
            Err(e) => {
                tracing::warn!("Found error: {:?}, retrying attempt #{}", e, attempt);
                tokio::time::sleep(Duration::from_secs(attempt as u64 + 1)).await;
            }
        }
        attempt += 1;
    }
}
