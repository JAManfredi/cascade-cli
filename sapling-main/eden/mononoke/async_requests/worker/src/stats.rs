/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::HashMap;
use std::time::Duration;

use async_requests::AsyncMethodRequestQueue;
use async_requests::types::RequestStatus;
use context::CoreContext;
use mononoke_api::RepositoryId;
use mononoke_types::Timestamp;
use requests_table::QueueStatsEntry;
use slog::warn;
use stats::define_stats;
use stats::prelude::*;

const STATS_LOOP_INTERNAL: Duration = Duration::from_secs(5 * 60);

const STATUSES: [RequestStatus; 4] = [
    RequestStatus::New,
    RequestStatus::InProgress,
    RequestStatus::Ready,
    RequestStatus::Polled,
];

define_stats! {
    prefix = "async_requests.worker.stats";

    stats_error: timeseries("error"; Count),
    queue_length_by_status: dynamic_singleton_counter("queue.length.{}", (status: String)),
    queue_age_by_status: dynamic_singleton_counter("queue.age_s.{}", (status: String)),
    queue_length_by_repo_and_status: dynamic_singleton_counter("queue.{}.length.{}", (repo_id: String, status: String)),
    queue_age_by_repo_and_status: dynamic_singleton_counter("queue.{}.age_s.{}", (repo_id: String, status: String)),
}

pub(crate) async fn stats_loop(
    ctx: &CoreContext,
    repo_ids: Vec<RepositoryId>,
    queue: &AsyncMethodRequestQueue,
) {
    loop {
        let res = queue.get_queue_stats(ctx).await;
        let now = Timestamp::now();
        match res {
            Ok(res) => {
                process_queue_length_by_status(ctx, &res);
                process_queue_age_by_status(ctx, now, &res);
                process_queue_length_by_repo_and_status(ctx, &repo_ids, &res);
                process_queue_age_by_repo_and_status(ctx, &repo_ids, now, &res);
            }
            Err(err) => {
                STATS::stats_error.add_value(1);
                warn!(
                    ctx.logger(),
                    "error while getting queue stats, skipping: {:?}", err
                );
            }
        }

        tokio::time::sleep(STATS_LOOP_INTERNAL).await;
    }
}

// Keep track of the stats we have already logged. Any missing ones, we will log a 0.
struct Seen {
    inner: HashMap<QueueStatsEntry, bool>,
}

impl Seen {
    fn new(repo_ids: &Vec<RepositoryId>) -> Self {
        let mut seen = HashMap::new();
        for status in STATUSES {
            for repo_id in repo_ids {
                seen.insert(
                    QueueStatsEntry {
                        repo_id: Some(repo_id.clone()),
                        status,
                    },
                    false,
                );
            }
        }
        Self { inner: seen }
    }

    fn mark(&mut self, repo_id: Option<RepositoryId>, status: RequestStatus) {
        self.inner.insert(QueueStatsEntry { repo_id, status }, true);
    }

    fn get_missing(&self) -> Vec<QueueStatsEntry> {
        self.inner
            .iter()
            .filter_map(|(entry, seen)| if !*seen { Some(entry) } else { None })
            .cloned()
            .collect()
    }
}

fn process_queue_length_by_status(ctx: &CoreContext, res: &requests_table::QueueStats) {
    let mut seen = Seen::new(&vec![]);
    let stats = &res.queue_length_by_status;
    for (status, count) in stats {
        seen.mark(None, *status);
        STATS::queue_length_by_status.set_value(ctx.fb, *count as i64, (status.to_string(),));
    }

    for entry in seen.get_missing() {
        STATS::queue_length_by_status.set_value(ctx.fb, 0, (entry.status.to_string(),));
    }
}

fn process_queue_age_by_status(
    ctx: &CoreContext,
    now: Timestamp,
    res: &requests_table::QueueStats,
) {
    let mut seen = Seen::new(&vec![]);
    let stats = &res.queue_age_by_status;
    for (status, ts) in stats {
        seen.mark(None, *status);
        let diff = std::cmp::max(now.timestamp_seconds() - ts.timestamp_seconds(), 0);
        STATS::queue_age_by_status.set_value(ctx.fb, diff, (status.to_string(),));
    }

    for entry in seen.get_missing() {
        STATS::queue_age_by_status.set_value(ctx.fb, 0, (entry.status.to_string(),));
    }
}

fn process_queue_length_by_repo_and_status(
    ctx: &CoreContext,
    repo_ids: &Vec<RepositoryId>,
    res: &requests_table::QueueStats,
) {
    let mut seen = Seen::new(repo_ids);
    let stats = &res.queue_length_by_repo_and_status;
    for (entry, count) in stats {
        seen.mark(entry.repo_id, entry.status);
        STATS::queue_length_by_repo_and_status.set_value(
            ctx.fb,
            *count as i64,
            (
                entry.repo_id.unwrap_or(RepositoryId::new(0)).to_string(),
                entry.status.to_string(),
            ),
        );
    }

    for entry in seen.get_missing() {
        STATS::queue_length_by_repo_and_status.set_value(
            ctx.fb,
            0,
            (
                entry.repo_id.unwrap_or(RepositoryId::new(0)).to_string(),
                entry.status.to_string(),
            ),
        );
    }
}

fn process_queue_age_by_repo_and_status(
    ctx: &CoreContext,
    repo_ids: &Vec<RepositoryId>,
    now: Timestamp,
    res: &requests_table::QueueStats,
) {
    let mut seen = Seen::new(repo_ids);
    let stats = &res.queue_age_by_repo_and_status;
    for (entry, ts) in stats {
        seen.mark(entry.repo_id, entry.status);
        let diff = std::cmp::max(now.timestamp_seconds() - ts.timestamp_seconds(), 0);
        STATS::queue_age_by_repo_and_status.set_value(
            ctx.fb,
            diff,
            (
                entry.repo_id.unwrap_or(RepositoryId::new(0)).to_string(),
                entry.status.to_string(),
            ),
        );
    }

    for entry in seen.get_missing() {
        STATS::queue_age_by_repo_and_status.set_value(
            ctx.fb,
            0,
            (
                entry.repo_id.unwrap_or(RepositoryId::new(0)).to_string(),
                entry.status.to_string(),
            ),
        );
    }
}
