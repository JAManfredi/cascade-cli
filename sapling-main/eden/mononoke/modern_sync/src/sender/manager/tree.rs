/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::collections::VecDeque;
use std::sync::Arc;
use std::sync::atomic::AtomicBool;
use std::sync::atomic::Ordering;
use std::time::Duration;

use anyhow::Result;
use context::CoreContext;
use futures::channel::oneshot;
use mercurial_types::HgManifestId;
use metaconfig_types::ModernSyncChannelConfig;
use mononoke_macros::mononoke;
use stats::define_stats;
use stats::prelude::*;
use tokio::sync::mpsc;
use tokio::time::interval;

use crate::sender::edenapi::EdenapiSender;
use crate::sender::manager::Manager;
use crate::sender::manager::TreeMessage;

define_stats! {
    prefix = "mononoke.modern_sync.manager.tree";

    synced_trees:  dynamic_timeseries("{}.synced_trees", (repo: String); Sum),
    content_wait_time_s:  dynamic_timeseries("{}.content_wait_time_s", (repo: String); Average),

    trees_queue_capacity: dynamic_singleton_counter("{}.trees.queue_capacity", (repo: String)),
    trees_queue_len: dynamic_histogram("{}.trees.queue_len", (repo: String); 10, 0,  100_000, Average; P 50; P 75; P 95; P 99),
    trees_queue_max_capacity: dynamic_singleton_counter("{}.trees.queue_max_capacity", (repo: String)),
}

pub(crate) struct TreeManager {
    config: ModernSyncChannelConfig,
    trees_recv: mpsc::Receiver<TreeMessage>,
}

impl TreeManager {
    pub(crate) fn new(
        config: ModernSyncChannelConfig,
        trees_recv: mpsc::Receiver<TreeMessage>,
    ) -> Self {
        Self { config, trees_recv }
    }

    async fn flush_trees(
        trees_es: &Arc<dyn EdenapiSender + Send + Sync>,
        batch_trees: &mut Vec<HgManifestId>,
        batch_done_senders: &mut VecDeque<oneshot::Sender<Result<()>>>,
        encountered_error: &mut Option<anyhow::Error>,
        reponame: &str,
    ) -> Result<(), anyhow::Error> {
        if !batch_trees.is_empty() || !batch_done_senders.is_empty() {
            let batch_size = batch_trees.len() as i64;
            if let Some(e) = encountered_error {
                let msg = format!("Error processing trees: {:?}", e);
                while let Some(sender) = batch_done_senders.pop_front() {
                    let _ = sender.send(Err(anyhow::anyhow!(msg.clone())));
                }
                tracing::error!("Error processing files/trees: {:?}", e);
                return Err(anyhow::anyhow!(msg.clone()));
            }

            if !batch_trees.is_empty() {
                let start = std::time::Instant::now();
                if let Err(e) = trees_es.upload_trees(std::mem::take(batch_trees)).await {
                    tracing::error!("Failed to upload trees: {:?}", e);
                    return Err(e);
                } else {
                    tracing::info!(
                        "Uploaded {} trees in {}ms",
                        batch_size,
                        start.elapsed().as_millis(),
                    );
                    STATS::synced_trees.add_value(batch_size, (reponame.to_owned(),));
                }
            }

            while let Some(sender) = batch_done_senders.pop_front() {
                let res = sender.send(Ok(()));
                if let Err(e) = res {
                    let msg = format!("Error sending content ready: {:?}", e);
                    tracing::error!("{}", msg);
                    return Err(anyhow::anyhow!(msg));
                }
            }
        }
        Ok(())
    }
}

impl Manager for TreeManager {
    fn start(
        mut self,
        ctx: CoreContext,
        reponame: String,
        trees_es: Arc<dyn EdenapiSender + Send + Sync>,
        cancellation_requested: Arc<AtomicBool>,
    ) {
        mononoke::spawn_task(async move {
            let trees_recv = &mut self.trees_recv;

            let mut encountered_error: Option<anyhow::Error> = None;
            let mut batch_trees = Vec::new();
            let mut batch_done_senders = VecDeque::new();
            let mut timer = interval(Duration::from_millis(self.config.flush_interval_ms as u64));

            while !cancellation_requested.load(Ordering::Relaxed) {
                tokio::select! {
                    msg = trees_recv.recv() => {
                        tracing::debug!("Trees channel capacity: {} max capacity: {} in queue: {}", trees_recv.capacity(), self.config.channel_size,  trees_recv.len());
                        STATS::trees_queue_capacity.set_value(ctx.fb, trees_recv.capacity() as i64, (reponame.clone(),));
                        STATS::trees_queue_len.add_value(trees_recv.len() as i64, (reponame.clone(),));
                        STATS::trees_queue_max_capacity.set_value(ctx.fb, trees_recv.max_capacity() as i64, (reponame.clone(),));
                        match msg {
                            Some(TreeMessage::WaitForContents(receiver)) => {
                                // Read outcome from content upload
                                let start = std::time::Instant::now();
                                match receiver.await {
                                    Ok(Err(e)) => {
                                        encountered_error.get_or_insert(e.context(
                                            "Contents error received. Winding down trees sender.",
                                        ));
                                    }
                                    Err(e) => {
                                        encountered_error.get_or_insert(anyhow::anyhow!(format!(
                                            "Error waiting for contents: {:#}",
                                            e
                                        )));
                                    }
                                    _ => (),
                                }
                                let elapsed = start.elapsed().as_secs();
                                STATS::content_wait_time_s.add_value(elapsed as i64, (reponame.clone(),));
                            }
                            Some(TreeMessage::Tree(t)) if encountered_error.is_none() => {
                                batch_trees.push(t);
                            }
                            Some(TreeMessage::TreesDone(sender)) => {
                                batch_done_senders.push_back(sender);
                            }
                            Some(TreeMessage::Tree(_)) => (),
                            None => break,
                        }
                        if batch_trees.len() >= self.config.batch_size as usize {
                            if let Err(e) = TreeManager::flush_trees(&trees_es, &mut batch_trees, &mut batch_done_senders, &mut encountered_error, &reponame).await {
                                tracing::error!("Trees flush failed: {:?}", e);
                                return;
                            }
                        }
                    }
                    _ = timer.tick() => {
                        if let Err(e) = TreeManager::flush_trees(&trees_es, &mut batch_trees, &mut batch_done_senders, &mut encountered_error, &reponame).await {
                            tracing::error!("Trees flush failed: {:?}", e);
                            return;
                        }
                    }
                }
            }
        });
    }
}
