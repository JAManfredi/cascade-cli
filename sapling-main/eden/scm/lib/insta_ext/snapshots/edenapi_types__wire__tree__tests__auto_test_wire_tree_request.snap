---
source: tree.rs
expression: "WireTreeRequest::arbitrary(&mut g)"

---
{"0":"0"}
